File: experiments/runs/2026-09-16_16-45-11_claim-office-example-mapping_exact-sol-v1.6-test-list-dimensions-cc_opus-5-no-thinking/src/price-list.ts
export interface Item {
  type: string;
  material?: string;
  enchantment?: number;
  cursed?: boolean;
}

interface PriceListEntry {
  insuranceValue: number;
  basePremium: number;
  isComponent: boolean;
}

const PRICE_LIST: Record<string, PriceListEntry> = {
  sword: { insuranceValue: 1000, basePremium: 100, isComponent: false },
  amulet: { insuranceValue: 600, basePremium: 60, isComponent: false },
  staff: { insuranceValue: 800, basePremium: 80, isComponent: false },
  potion: { insuranceValue: 400, basePremium: 40, isComponent: false },
  rune: { insuranceValue: 250, basePremium: 25, isComponent: true },
  moonstone: { insuranceValue: 250, basePremium: 25, isComponent: true },
};

const BLOCK_SIZE = 3;
const BLOCK_BASE_PREMIUM = 60;

function entryFor(type: string): PriceListEntry {
  const entry = PRICE_LIST[type];
  if (entry === undefined) {
    throw new Error(`The MHPCO does not insure items of type "${type}"`);
  }
  return entry;
}

export function insuranceValueOf(item: Item): number {
  return entryFor(item.type).insuranceValue;
}

export function basePremiumOf(item: Item): number {
  return entryFor(item.type).basePremium;
}

/**
 * The base premium for a group of alike items -- items of the very same type.
 * A building block of exactly 3 alike components is offered at a special price.
 */
/** A building block is exactly 3 alike components -- main items never qualify. */
function formsBuildingBlock(type: string, count: number): boolean {
  return entryFor(type).isComponent && count === BLOCK_SIZE;
}

export function basePremiumForAlike(type: string, count: number): number {
  return formsBuildingBlock(type, count)
    ? BLOCK_BASE_PREMIUM
    : count * entryFor(type).basePremium;
}

File: experiments/runs/2026-09-20_07-33-00_claim-office-java-example-mapping_exact-ptdd-v1-cc_opus-5-no-thinking/src/main/java/PriceList.java
import java.util.Map;
import java.util.Set;

/**
 * The MHPCO price list: the insurance value and base premium of every item
 * type the office covers. Components are insured at a uniform value, with a
 * special base premium for a building block of 3 alike components.
 */
final class PriceList {

    private static final Set<String> COMPONENT_TYPES = Set.of("rune", "moonstone");
    private static final double COMPONENT_BASE_PREMIUM = 25;
    private static final int COMPONENT_INSURANCE_VALUE = 250;
    private static final int BLOCK_SIZE = 3;
    private static final double BLOCK_BASE_PREMIUM = 60;

    private static final Map<String, Integer> MAIN_ITEM_INSURANCE_VALUES = Map.of(
            "sword", 1000,
            "amulet", 600,
            "staff", 800,
            "potion", 400);

    private static final Map<String, Double> MAIN_ITEM_BASE_PREMIUMS = Map.of(
            "sword", 100.0,
            "amulet", 60.0,
            "staff", 80.0,
            "potion", 40.0);

    private PriceList() {
    }

    static boolean isComponent(String type) {
        return COMPONENT_TYPES.contains(type);
    }

    static double basePremiumOf(Item item) {
        if (isComponent(item.type())) {
            return COMPONENT_BASE_PREMIUM;
        }
        return requireCovered(MAIN_ITEM_BASE_PREMIUMS.get(item.type()), item.type());
    }

    static int insuranceValueOf(Item item) {
        if (isComponent(item.type())) {
            return COMPONENT_INSURANCE_VALUE;
        }
        return requireCovered(MAIN_ITEM_INSURANCE_VALUES.get(item.type()), item.type());
    }

    /** A building block of exactly 3 alike components is offered at a special premium. */
    static double basePremiumOfAlikeComponents(String type, int count) {
        if (count == BLOCK_SIZE) {
            return BLOCK_BASE_PREMIUM;
        }
        return count * COMPONENT_BASE_PREMIUM;
    }

    private static <T> T requireCovered(T price, String type) {
        if (price == null) {
            throw new IllegalArgumentException("the MHPCO does not cover items of type: " + type);
        }
        return price;
    }
}
